### ISSUE TITLE
Magento 2.2.7 custom product attributes don't load when switching between attribute sets on product edit

### ISSUE THREAD

**Issue by Lez-2** [magento/magento2#20997]
### Preconditions (*)
1. Magento 2.2.7

### Steps to reproduce (*)
1. Have a minimum of 2 product attribute sets which have at least 1 different attribute inside of them (for example, attribute set: "Default" and "Default + size" - where in the "Default + size" set an additional attribute called "Size" resides
2. Create a product with the "Default" attribute set
3. Edit the same product and change the attribute set to "Default + size"
4. You will notice that a loader starts, but it doesn't add the "Size" attribute to the available fields.

### Expected result (*)
1. The custom attribute (in the above example "Size") should show

### Actual result (*)
1. The custom attribute (in the above example "Size") doesn't show
2. When you save the product with the new attribute set then the custom attribute shows

### Additional info
While creating a new product and selecting the attribute set with the additional custom attributes this works as expected (all product attributes are shown/loaded)
This works as expected in Magento 2.2.2 - 2.2.4

**Comment by magento-engcom-team** [comment]
Hi @Lez-2. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@Lez-2 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by shikhamis11** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11, here is your [Magento instance](https://i-20997-2-2-develop.instances.magento-community.engineering).
Admin access: [https://i-20997-2-2-develop.instances.magento-community.engineering/admin](https://i-20997-2-2-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by shikhamis11** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11, here is your [Magento instance](https://i-20997-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-20997-2-3-develop.instances.magento-community.engineering/admin](https://i-20997-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @shikhamis11 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98100`, `MAGETWO-98101` were created

**Issue Available**: @shikhamis11, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if you want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by shikhamis11** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11, here is your [Magento instance](https://i-20997-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-20997-2-3-develop.instances.magento-community.engineering/admin](https://i-20997-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by girish-codilar** [comment]
@magento-engcom-team give me 2.2.7 instance

**Comment by magento-engcom-team** [comment]
Hi @girish-codilar. Thank you for your request. I'm working on Magento 2.2.7 instance for you

**Comment by magento-engcom-team** [comment]
Hi @girish-codilar, here is your [Magento instance](https://i-20997-2-2-7.instances.magento-community.engineering).
Admin access: [https://i-20997-2-2-7.instances.magento-community.engineering/admin](https://i-20997-2-2-7.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by Ctucker9233** [comment]
@magento-engcom-team I confirm the same behavior on 2.2.8. An annoying work around is to save the product and then the custom attribute field appears. Would appreciate an update on this issue.

**Comment by engcom-Charlie** [comment]
Hello, @Lez-2 we aren't able to reproduce the issue on the latest 2.4-develop by following steps:

1. Create a new attribute set for example `AS`.
2. Assign new attribute to attribute set `AS` (i used default `Manufacturer` which isn't assigned to `Default` attribute set).
3. Create a product on the `Default` attribute set.
4. Open this product for edit.
5. Switch to `AS` attribute set.

**AR**: Manufacturer attribute is visible as expected.
Looks like the issue was fixed here https://github.com/magento/magento2/commit/bcf7d9c0ef8cbd514774c24017e3d83a60353538

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.4-develop` branch. Thank you for your collaboration.

Thanks for your report!

**Comment by magento-engcom-team** [comment]
Hi @Lez-2, @engcom-Charlie. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-24144` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
